### ISSUE TITLE
Why does mainloop quit, when core_event::done is fired?

### ISSUE THREAD

**Issue by mincequi** [Curve/rohrkabel#10]
Hi there and thanks for this nice pipewire wrapper.

I have a question:
consider this code snippet:

```
int main() {
    auto main_loop = pw::main_loop::create();
    auto context   = pw::context::create(main_loop);
    auto core      = pw::core::create(context);
    auto reg       = pw::registry::create(core);
    auto listener  = reg->listen();

    listener.on<pipewire::registry_event::global>([&](const pw::global& global) {
        if (global.type == pw::node::type) {
            auto node = *reg->bind<pw::node>(global.id).get();
            info("node: {}", node.id());
        }
    });

    main_loop->run();
    return 0;
}

```

This will run and quit as soon as all nodes are printed. Why is this happening? Actually, it should continue and print all nodes that are subsequently added.

Thanks in advance!

**Comment by Curve** [comment]
Hi, thanks for the kind words!

The default overload for bind issues a core update, which in turn calls quit. You can prevent this by passing `update_strategey::none` as the second Parameter to bind.

**Comment by mincequi** [comment]
Ok, i guess i understand your ideas. However, i believe that the implemented behavior is a little inconsistent:
the mainloop will run forever if i listen to global events (like ports, nodes, links), but it stops as soon as i received all properties from a bound object (metadata). My assumption is, that it keeps running, until i quit it explicitly.
But this rises the question what to do if i want to integrate the pipewire mainloop into another mainloop implementation...?

To come back to your suggestion:
yes, i believe the callback based would fit better here (and also to my implementation, where i wrap all callback into reactive streams): https://github.com/mincequi/KlangDienst

Thanks already!

**Comment by Curve** [comment]
> Ok, i guess i understand your ideas. However, i believe that the implemented behavior is a little inconsistent: the mainloop will run forever if i listen to global events (like ports, nodes, links), but it stops as soon as i received all properties from a bound object (metadata). 

Yes - This is what happens when you use the default bind overload. As mentioned previously, you can either pass `update_strategy::none` to avoid the sync, but with the draw-back that you'll have to make sure the object is available yourself when you try to access it later.

Also, internally bind only uses the already existing listener interface - You can also fallback to that for the time being :)

> My assumption is, that it keeps running, until i quit it explicitly. But this rises the question what to do if i want to integrate the pipewire mainloop into another mainloop implementation...?

I'm not sure how the current behavior impacts this, in the examples a simple while loop is used to run the main-loop as long as required, you could simply set the loop condition to false and then quit the main-loop explicitly

**Comment by Curve** [comment]
Been quite busy lately but I'll look into this soon! I'm currently considering adding coroutine support as that seems like a good fit :)

**Comment by Curve** [comment]
I'm happy to announce that [v9.0.0](https://github.com/Curve/rohrkabel/releases/tag/v9.0.0) has just been released :)

I will mark this a fixed ^^
